File: app/src/main/java/com/udangtangtang/haveibeen/util/PictureScanHelper.java
package com.udangtangtang.haveibeen.util;

import android.content.Context;
import android.database.Cursor;

import androidx.exifinterface.media.ExifInterface;

import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.os.Build;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import com.udangtangtang.haveibeen.model.DBHelper;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;

public class PictureScanHelper {
    private final String TAG = "pictureManager";
    private Context context;
    private ExifInterface exifInterface;
    private GeocodingHelper geocodingHelper;

    public PictureScanHelper(Context mContext) {
        context = mContext;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
        /*
        /  사진 파일 리스트
        /  https://3001ssw.tistory.com/195
        */
    public ArrayList<String> scanPictures(Context context) {
        ArrayList<String> fileList = new ArrayList<>();
        Uri uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
        String[] projection = {MediaStore.MediaColumns.DATA, MediaStore.MediaColumns.DISPLAY_NAME};

        Cursor cursor = context.getContentResolver().query(uri, projection, null, null, MediaStore.MediaColumns.DATE_ADDED + " desc");
        int columnIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA);
        int columnDisplayName = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME);
        int lastIndex;

        DBHelper dbHelper = new DBHelper(context);
        SQLiteDatabase sqlDB = dbHelper.getReadableDatabase();
        while (cursor.moveToNext()) {
            String absolutePathOfImage = cursor.getString(columnIndex);
            String nameOfFile = cursor.getString(columnDisplayName);
            lastIndex = absolutePathOfImage.lastIndexOf(nameOfFile);
            lastIndex = lastIndex >= 0 ? lastIndex : nameOfFile.length() - 1;

            if (!TextUtils.isEmpty(absolutePathOfImage)) {
                // 이미 추가된 사진인지 확인
                String[] params = new String[]{absolutePathOfImage};
                Cursor cursor1 = sqlDB.rawQuery("select filename from myDB where filename=?;", params);
                if (cursor1 != null && cursor1.moveToFirst()) {
                    // 존재하면 pass
                } else {
                    // 존재하지 않으면 추가
                    fileList.add(absolutePathOfImage);
                }
            }
        }
        sqlDB.close();

        return fileList;
    }
    ////////////////////////////////////////////////////////////////////////////////////////////////

    public void initializePictureDB(DBHelper dbHelper, ArrayList<String> fileList) {
        double[] latLong = new double[2];
        int newPictureCount = 0;
        int noLatLngCount = 0;
        String address = "";
        String datetime = "";
        for (int i = 0; i < fileList.size(); i++) {
            try {
                // 위/경도, 일시, 주소 얻기
                if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.Q){
                    Uri photoUri = MediaStore.setRequireOriginal(Uri.parse(fileList.get(i)));
                    InputStream stream = context.getContentResolver().openInputStream(photoUri);
                    if (stream != null) {
                        ExifInterface exifInterface = new ExifInterface(stream);
                        latLong = exifInterface.getLatLong();
                        datetime = exifInterface.getAttribute(android.media.ExifInterface.TAG_DATETIME);
                        stream.close();
                    } else {
                        latLong = new double[2];
                        datetime = exifInterface.getAttribute(android.media.ExifInterface.TAG_DATETIME);
                    }
                }else{
                    exifInterface = new ExifInterface(fileList.get(i));
                    latLong = exifInterface.getLatLong();
                    datetime = exifInterface.getAttribute(android.media.ExifInterface.TAG_DATETIME);
                }
            } catch (IOException e) {
                System.out.println(e.toString());
            }
            try {
                geocodingHelper = new GeocodingHelper(context, latLong[0], latLong[1]);
                address = geocodingHelper.getAddress();
                dbHelper.insertImageToDB(fileList.get(i), latLong[0], latLong[1], address, datetime);
                Log.i(TAG, "Image " + fileList.get(i) + address);
                newPictureCount += 1;
            } catch (NullPointerException e) {
                // 좌표 정보가 없을 경우
                Log.i(TAG, "Image " + fileList.get(i) + " has no coordination info.");
                noLatLngCount += 1;
            }
        }

        // 탐색 결과 출력
        if (newPictureCount != 0) {
            Toast.makeText(context, "새로운 사진 " + newPictureCount + "장을 찾았습니다.", Toast.LENGTH_LONG).show();
        }

        if (noLatLngCount != 0) {
            Toast.makeText(context, "좌표 정보가 없는 사진 " + noLatLngCount + "장을 찾았습니다. \n지도에는 추가되지 않습니다.", Toast.LENGTH_LONG).show();
        }

        if(newPictureCount==0&&noLatLngCount==0){
            Toast.makeText(context, "새로운 사진이 없습니다.", Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/com/udangtangtang/haveibeen/util/PictureScanHelper.kt
package com.udangtangtang.haveibeen.util

import android.app.Application
import android.content.*
import android.icu.text.SimpleDateFormat
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import android.text.TextUtils
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.exifinterface.media.ExifInterface
import com.udangtangtang.haveibeen.activity.MainActivity
import com.udangtangtang.haveibeen.entity.PictureEntity
import com.udangtangtang.haveibeen.repository.RecordRepository
import kotlinx.coroutines.*
import okhttp3.Dispatcher
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

@RequiresApi(Build.VERSION_CODES.TIRAMISU)
class PictureScanHelper(private val context: Context, private val db:RecordRepository) {
    // TODO : 미디어 저장소 업데이트 감지, 동영상 스캔 추가, AsyncTask->Coroutine, 이전 버전?
    // https://developer.android.com/training/data-storage/shared/media?hl=ko#detect-updates-media-files
    private val TAG = "PictureScanHelper"
    private var geocodingHelper: GeocodingHelper

    init{
        geocodingHelper= GeocodingHelper(context, db)
    }

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    fun scanPictures() {
        val uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
        val projection = arrayOf(MediaStore.MediaColumns._ID,
                                MediaStore.MediaColumns.DATA,
                                MediaStore.MediaColumns.DISPLAY_NAME,
                                MediaStore.MediaColumns.DATE_ADDED)
        val selectionArgs = arrayOf(
            dateToTimestamp(day = 1, month = 1, year = 1970).toString())
        val cursor = context.contentResolver.query(
            uri,
            projection,
            null,
            null,
            MediaStore.MediaColumns.DATE_TAKEN + " desc"
        )?.use{ cursor->
            var pictureList = mutableListOf<PictureEntity>()
            Log.d(TAG, cursor.count.toString())
            while (cursor.moveToNext()) {
                val idColumnIndex=cursor.getColumnIndexOrThrow(MediaStore.MediaColumns._ID)
                val absolutePathOfImage =
                    cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA))
                val nameOfFile =
                    cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME))
                var datetime=
                    cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATE_ADDED))
                val pictureAmount=cursor.count.toFloat()
                if (!TextUtils.isEmpty(absolutePathOfImage)) {
                    // Update Progress
                    (context as MainActivity).updateProgress((cursor.position+1).toFloat()/pictureAmount)
                    Log.d(TAG, "curr : "+cursor.position.toString()+"/"+pictureAmount.toString())
                    // 위/경도, 일시, 주소 얻기
                    var latLong: DoubleArray? = null
                    datetime=DateTimeFormatter.ofPattern("yyyy년 M월 dd일 HH:mm").format(
                        LocalDateTime.ofInstant(Instant.ofEpochMilli(datetime.toLong() * 1000), ZoneOffset.UTC)
                    )
                    var photoUri = Uri.withAppendedPath(
                        MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                        cursor.getString(idColumnIndex)
                    )
                    photoUri = MediaStore.setRequireOriginal(photoUri)
                    context.contentResolver.openInputStream(photoUri)?.use { stream ->
                        val exifInterface=ExifInterface(stream).run {latLong = this.latLong}
                    }

                    // 좌표 정보를 통해 주소를 얻고 이를 파일명, 위/경도 정보, 촬영 일시 등을 모두 DB에 추가
                    CoroutineScope(Dispatchers.IO).launch {
                        async {
                            if (latLong?.isNotEmpty() == true and !db.isExistPicture(
                                    latLong!![0],
                                    latLong!![1],
                                    nameOfFile
                                )
                            ) {
                                val picture = PictureEntity(
                                    latLong!![0],
                                    latLong!![1],
                                    absolutePathOfImage,
                                    nameOfFile,
                                    null,
                                    datetime
                                )
                                pictureList.add(picture)
                                db.addPicture(picture)
                            } else {
                                // 좌표 정보가 없을 경우
                                Log.d(
                                    TAG,
                                    "Image " + absolutePathOfImage + " has no coordination info."
                                )
                            }
                        }.await()
                        async {
                            geocodingHelper.getAddress(latLong!![0], latLong!![1])
                        }
                    }

                }
            }
        }
    }

    private fun dateToTimestamp(day: Int, month: Int, year: Int): Long =
        SimpleDateFormat("yyyy.MM.dd").let { formatter ->
            formatter.parse("$year.$month.$day")?.time ?: 0
        }

    ////////////////////////////////////////////////////////////////////////////////////////////////
//    fun initializePictureDB(dbHelper: DBHelper, fileList: ArrayList<String>) {
//        var latLong: DoubleArray? = DoubleArray(2)
//        var newPictureCount = 0
//        var noLatLngCount = 0
//        var address: String? = ""
//        var datetime: String? = ""

//        }
//
//        // 탐색 결과 출력
//        if (newPictureCount != 0) {
//            Toast.makeText(context, "새로운 사진 " + newPictureCount + "장을 찾았습니다.", Toast.LENGTH_LONG)
//                .show()
//        }
//        if (noLatLngCount != 0) {
//            Toast.makeText(
//                context, """
//     좌표 정보가 없는 사진 ${noLatLngCount}장을 찾았습니다.
//     지도에는 추가되지 않습니다.
//     """.trimIndent(), Toast.LENGTH_LONG
//            ).show()
//        }
//        if (newPictureCount == 0 && noLatLngCount == 0) {
//            Toast.makeText(context, "새로운 사진이 없습니다.", Toast.LENGTH_LONG).show()
//        }
//    }


}